source/server/connection_handler_impl.h

```diff
@@ -102,51 +97,66 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
   };
 
 private:
-  struct ActiveConnection;
-  using ActiveConnectionPtr = std::unique_ptr<ActiveConnection>;
-  struct ActiveSocket;
-  using ActiveSocketPtr = std::unique_ptr<ActiveSocket>;
+  struct ActiveTcpConnection;
+  using ActiveTcpConnectionPtr = std::unique_ptr<ActiveTcpConnection>;
+  struct ActiveTcpSocket;
+  using ActiveTcpSocketPtr = std::unique_ptr<ActiveTcpSocket>;
 
   /**
    * Wrapper for an active tcp listener owned by this handler.
    */
-  class ActiveTcpListener : public Network::ListenerCallbacks, public ActiveListenerImplBase {
+  class ActiveTcpListener : public Network::ListenerCallbacks,
+                            public ActiveListenerImplBase,
+                            public Network::BalancedConnectionHandler {
   public:
     ActiveTcpListener(ConnectionHandlerImpl& parent, Network::ListenerConfig& config);
     ActiveTcpListener(ConnectionHandlerImpl& parent, Network::ListenerPtr&& listener,
                       Network::ListenerConfig& config);
     ~ActiveTcpListener() override;
+    void onAcceptWorker(Network::ConnectionSocketPtr&& socket,
+                        bool hand_off_restored_destination_connections, bool rebalanced);
+    void decNumConnections() {
+      ASSERT(num_listener_connections_ > 0);
+      num_listener_connections_--;
+    }
 
     // Network::ListenerCallbacks
-    void onAccept(Network::ConnectionSocketPtr&& socket,
-                  bool hand_off_restored_destination_connections) override;
-    void onNewConnection(Network::ConnectionPtr&& new_connection) override;
+    void onAccept(Network::ConnectionSocketPtr&& socket) override;
+
+    // Network::BalancedConnectionHandler
+    uint64_t numConnections() const override { return num_listener_connections_; }
+    void incNumConnections() override { num_listener_connections_++; }
```

Comment:
I'm fairly certain the compiler will generate the same code in either case. If I change this here I will change things elsewhere to be consistent but I'm not sure it's worth it? Any particular reason to change it?

Reply:
It might optimize to the same thing, but since this is an atomic, I was guessing that  for the ++ override, the compiler would generate something like:
```
  acquire_lock
  temp = value;
  ++value;
  release_lock
  return value
```
then after inlining I could see the compiler might discover that the returned value was not used, and thus the temp was dead, and get the same thing in the end.

In the distant past I've always been surprised at things the compiler did, both beyond my expectations, and where it would not do things I might expect it to be able to do.

So I always got into the habit of using ++x rather than x++ unless I really needed the post-increment behavior.

FWIW the public Google style guide https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement says this:
> For simple scalar (non-object) values there is no reason to prefer one form and we allow either. For iterators and other template types, use pre-increment.
and this is a template type so I guess the style guide says make the change :)

You may be right; it might be the same in the end. But it's never worse to use ++x..

Reply:
(note: edited two typos after sending....)

Reply:
OK I will change it in this cc/h though we should try to be consistent about this elsewhere probably.

Reply:
(Later I will look at the disassembled code also to verify it generates the same code out of curiosity. Given that it compiles down to a `lock inc` instruction I'm pretty sure it will.)

Reply:
Thanks I'm curious too...

Reply:
@jmarantz @danzh2010 for posterity: 

```
000000000058f1f0 <_ZN5Envoy6Server21ConnectionHandlerImpl17ActiveTcpListener17incNumConnectionsEv>:
      { return __atomic_add_fetch(&_M_i, 1, memory_order_seq_cst); }
  58f1f0:       f0 48 81 87 c0 00 00    lock addq $0x1,0xc0(%rdi)
  58f1f7:       00 01 00 00 00 
    void incNumConnections() override { ++num_listener_connections_; }
  58f1fc:       c3                      retq

000000000058f1f0 <_ZN5Envoy6Server21ConnectionHandlerImpl17ActiveTcpListener17incNumConnectionsEv>:
      }

      _GLIBCXX_ALWAYS_INLINE __int_type
      fetch_add(__int_type __i,
                memory_order __m = memory_order_seq_cst) noexcept
      { return __atomic_fetch_add(&_M_i, __i, __m); }
  58f1f0:       f0 48 81 87 c0 00 00    lock addq $0x1,0xc0(%rdi)
  58f1f7:       00 01 00 00 00 
    void incNumConnections() override { num_listener_connections_++; }
  58f1fc:       c3                      retq
```

Reply:
very cool, the compiler cleans it all up. thanks for confirming!